File: 9020/9020.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    9020.py                                            :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2021/10/07 19:37:54 by jihoolee          #+#    #+#              #
#    Updated: 2021/10/07 19:48:52 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

def eratos(max):
	is_prime = [True] * (max + 1)
	for i in range(2, max + 1):
		if (is_prime[i] == True):
			for j in range(2 * i, max + 1, i):
				is_prime[j] = False
	return ([i for i in range(2, max) if is_prime[i] == True])

def get_goldbach(num):
	smaller = int(num / 2)
	while (smaller not in prime_nums_lst or num - smaller not in prime_nums_lst):
		smaller -= 1
	return (smaller, num - smaller)

cases = int(input())

prime_nums_lst = eratos(10000)
for _ in range(cases):
	 num = int(input())
	 print(*get_goldbach(num))

File: 9020/9020.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   9020.cpp                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2021/10/07 16:16:13 by jihoolee          #+#    #+#             */
/*   Updated: 2021/10/07 19:36:48 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <vector>

using namespace std;

vector<bool>	is_prime(10000 + 1, true);

void	eratos()
{
	int	max = is_prime.size() - 1;

	is_prime[0] = false;
	is_prime[1] = false;
	for (int i = 2; i * i <= max; i++)
		if (is_prime[i] == true)
			for (int j = i * 2; j <= max; j += i)
				is_prime[j] = false;
}

int	get_goldbach(int num)
{
	int	smaller = num / 2;

	while (is_prime[smaller] == false || is_prime[num - smaller] == false)
		smaller--;
	return (smaller);
}

int	main(void)
{
	int	cases;
	int	num;
	int	smaller_gb;

	cin >> cases;
	eratos();
	for (int i = 0; i < cases; i++)
	{
		cin >> num;
		smaller_gb = get_goldbach(num);
		cout << smaller_gb << ' ' << num - smaller_gb << '\n';
	}
	return (0);
}
